File: 2022/03/python/3a.py
def create_halves(rucksack):
    half_len = int(len(rucksack)/2)
    return rucksack[:half_len], rucksack[half_len:]

def find_shared_items(first_half, second_half):
    return set(first_half) & set(second_half)

def calculate_item_priority(letter):
    if letter.islower():
        return ord(letter) - ord("a") + 1
    else: 
        return ord(letter) - ord("A") + 27

def main(input_path):
    all_items = []
    with open(input_path) as f:
        for rucksack in f:
            first_half, second_half = create_halves(rucksack.strip())
            shared_items = find_shared_items(first_half, second_half)
            all_items.append(*shared_items)
    items_values = [calculate_item_priority(el) for el in all_items]
    return sum(items_values)

print(main("../input.txt"))

File: 2022/03/go/3a.go
package main

import (
	"strings"
)

func createHalves(rucksack string) (string, string) {
	var half_len int = len(rucksack) / 2
	return rucksack[:half_len], rucksack[half_len:]
}

func findIntersection(firstHalf string, secondHalf string, intersection map[string]int) {
	for _, letter := range firstHalf {
		letter_str := string(letter)

		if strings.Contains(secondHalf, letter_str) {
			if _, exists := intersection[letter_str]; !exists {
				intersection[letter_str] = calculateItemPriority(letter_str)
			}
		}
	}
}

func calculateItemPriority(letter string) int {
	letter_rune := []rune(letter)[0]
	if strings.ToLower(letter) == letter {
		return  int(letter_rune - 'a') + 1
	}
	return int(letter_rune - 'A') + 27
}

func solution_a(rucksack string, intersection map[string]int){

	first_half, second_half := createHalves(rucksack)
	findIntersection(first_half, second_half, intersection)
}